src/Sarif/FileRegionsCache.cs

```diff
@@ -47,24 +51,213 @@ public Region PopulatePrimaryRegionProperties(PhysicalLocation physicalLocation,
+
+            // Note: execution order of these helpers is important, as some 
+            // calls assume that certain preceding helpers have executed,
+            // with the result that certain properties are populated
+
+            // Populated at this point: StartLine
+            PopulateEndLine(lineIndex, region);
+
+            // Populated at this point: StartLine, EndLine
+            PopulateStartColumn(region);
+
+            // Populated at this point: StartLine, EndLine, StartColumn
+            PopulateEndColumn(lineIndex, region, fileText);
+
+            // Populated at this point: StartLine, EndLine, StartColumn, EndColumn
+            PopulateCharOffset(lineIndex, region, fileText);           
+
+            // Populated at this point: StartLine, EndLine, StartColumn, EndColumn, CharOffset
+            PopulateCharLength(lineIndex, region);
+
+            // Populated at this point: StartLine, EndLine, StartColumn, EndColumn, CharOffset, CharLength
+            Debug.Assert(region.StartLine > 0);
+            Debug.Assert(region.EndLine > 0);
+            Debug.Assert((region.CharOffset + region.CharLength) <= fileText.Length);
+            Debug.Assert(region.StartColumn > 0);
+            Debug.Assert(region.CharLength > 0 || (region.StartColumn == region.EndColumn && region.StartLine == region.EndLine));
+            Debug.Assert(region.EndColumn > 0);
+        }
+
+        private static void PopulateEndLine(NewLineIndex lineIndex, Region region)
+        {
+            // Populated at this point: StartLine
+            Debug.Assert(region.StartLine > 0);
+
+            region.EndLine = region.EndLine == 0 ? region.StartLine : region.EndLine;
+        }
+
+        private static void PopulateStartColumn(Region region)
+        {
+            // Populated at this point: StartLine, EndLine
+            Debug.Assert(region.StartLine > 0);
+            Debug.Assert(region.EndLine > 0);
+
+            region.StartColumn = region.StartColumn == 0 ? 1 : region.StartColumn;
+        }
+
+
+        private void PopulateEndColumn(NewLineIndex lineIndex, Region region, string fileText)
+        {
+            // Populated at this point: StartLine, EndLine, StartColumn
+            Debug.Assert(region.StartLine > 0);
+            Debug.Assert(region.StartColumn > 0);
+            Debug.Assert(region.EndLine > 0);
+
+            if (region.EndColumn == 0)
+            {
+                // No explicit end column. Increment from end line through
+                // the end of the line, excluding new line characters
+                LineInfo lineInfo = lineIndex.GetLineInfoForLine(region.EndLine);
+                int endColumnOffset = lineInfo.StartOffset;
+
+                while (endColumnOffset < fileText.Length &&
+                       !NewLineIndex.s_newLineCharSet.Contains(fileText[endColumnOffset]))
+                {
+                    endColumnOffset++;
+                }
+
+                // End columns are 1-indexed
```

Comment:
No, take another look. We are, in fact, pointing at the newLine character. the +1 is, in fact, because EndColumn is 1-indexed, while the offset data is 0-indexed.

---
In reply to: [204583183](https://github.com/Microsoft/sarif-sdk/pull/957#discussion_r204583183) [](ancestors = 204583183)